TF-M: Add secure peripheral configuration and sample

joerchan opened the pull request:
Add configuration to configure peripherals as secure in TF-M.
Pull in TF-M secure peripheral fixes for TF-M.
Add sample to demonstrate configuration of secure peripherals in secure partition defined by the application.

joerchan commented on samples/tfm/tfm_secure_peripheral/boards/nrf5340dk_nrf5340_cpuapp.conf:
```diff
@@ -0,0 +1,3 @@
+
+CONFIG_NRF_GPIO0_PIN_MASK_SECURE=0x00800000 # P0: 23
+CONFIG_NRF_GPIO1_PIN_MASK_SECURE=0x0000a000 # P1: 13, 15
```
@mbolivar-nordic Let me know if you can come up with something better than this.

mbolivar-nordic replied:
> @mbolivar-nordic Let me know if you can come up with something better than this.

Can you put these into devicetree as properties?

joerchan replied:
I don't know how to do that in a way that is accesible in Kconfig.

I need to provide this to a different build system so our options are limited by that, simplest is to provide everything in a simple define.

mbolivar-nordic replied:
> I don't know how to do that in a way that is accesible in Kconfig.

What do you mean? That's what many of the helpers in kconfigfunctions.py are for.

Note also that direct access to the devicetree in the build system is available in https://github.com/nrfconnect/sdk-zephyr/blob/bb39dfe3121a1c245d7c5804de0e14e19b299b83/cmake/modules/extensions.cmake#L2490

joerchan replied:
Mostly I mean that I'm not that skilled in DT to know how to do that.

This one seems a bit more complicated than most uses though, because I want a bitmask of all of the GPIOs, so I would need to aggregate many symbols then, wouldn't I?

Also I'm unsure how good idea this is since DPPI channels work the same way, and I don't recall seeing DPPI channels being used in DT, it would be best to have them be configured in the same way I think.

mbolivar-nordic replied:
> Mostly I mean that I'm not that skilled in DT to know how to do that.

I don't know the problem area so I may be missing something, but here is what I recommend:

https://github.com/mbolivar-nordic/sdk-zephyr/commit/ac6983bfdf2681cf7001fb9eb209cf051ce97a48

Then you can get the value out into Kconfig if you really need it there with kconfigfunctions. But I am not sure why you need it in Kconfig. It looks like you could do something like this to access the property directly instead:

```c
 #ifdef NRF_P0
-		PERIPH("NRF_P0", NRF_P0, CONFIG_SPM_NRF_P0_NS),
+		PERIPH("NRF_P0", NRF_P0, DT_PROP_OR(DT_NODELABEL(gpio0), secure_pin_mask, 0)),
 #endif
```

And similarly for port 1, just use `DT_NODELABEL(gpio1)` instead.

> This one seems a bit more complicated than most uses though, because I want a bitmask of all of the GPIOs, so I would need to aggregate many symbols then, wouldn't I?

> Also I'm unsure how good idea this is since DPPI channels work the same way, and I don't recall seeing DPPI channels being used in DT, it would be best to have them be configured in the same way I think.

I think that's likely just a historical accident; if we were configuring these today they'd probably belong in DT, I guess, since that's where boot-time hardware configuration is supposed to go.

joerchan replied:
Offline discussion: We will use Kconfig for now, switch to DT later when it supports the use-case better.

NordicBuilder commented:
You can find the documentation preview for this PR at this <a href='http://developer.nordicsemi.com/nRF_Connect_SDK_dev/doc/PR-7148
'>link</a>. It will be updated about 10 minutes after the documentation build succeeds.

*Note: This comment is automatically posted by the Documentation Publishing GitHub Action.*

tejlmand commented on subsys/nonsecure/Kconfig:
```diff
@@ -0,0 +1,314 @@
+	bool "FPU configured as secure peripheral"
+	depends on CPU_HAS_FPU
+
+config CONFIG_NRF_DCNF_SECURE
```
a `CONFIG_` too much ?
```suggestion
config NRF_DCNF_SECURE
```

tejlmand commented on subsys/nonsecure/Kconfig:
```diff
@@ -0,0 +1,314 @@
+menu "Trusted Execution Non-Secure"
+menu "Peripheral Secure mapping"
+
+config NRF_FPU_SECURE
+	bool "FPU configured as secure peripheral"
+	depends on CPU_HAS_FPU
```
General comment.

All settings in this file have no help text.

Are we sure all users fully understand what it implies to configure a peripheral as secure ?

Could we have just a brief description and maybe a reference to more information.
Remember the help text is rendered into our docs, like this:
http://developer.nordicsemi.com/nRF_Connect_SDK/doc/latest/kconfig/CONFIG_BUILD_WITH_TFM.html#std-kconfig-CONFIG_BUILD_WITH_TFM

tejlmand commented on modules/tfm/tfm/boards/board/tfm_peripherals_config.h:
```diff
@@ -11,18 +11,291 @@
+#include <autoconf.h>
+#include <nrfx.h>
+
+#if CONFIG_NRF_DCNF_SECURE
+#define TFM_PERIPHERAL_DCNF_SECURE                  1
+#endif
```
enabled bool configs are defined as `1`, and disabled Kconfig settings are not in `autoconf.h`, so this should work just as well:
```suggestion
#define TFM_PERIPHERAL_DCNF_SECURE                  CONFIG_NRF_DCNF_SECURE
```

could you try that ?
I think it looks cleaner.

tejlmand commented on modules/tfm/zephyr/Kconfig:
```diff
@@ -3,6 +3,7 @@
 #
 # SPDX-License-Identifier: LicenseRef-Nordic-5-Clause
 #
+menu "TFM"
```
why creating this menu when all settings inside the menu are promptless ?

joerchan replied:
Left-over from refactoring. I had some configs in here earlier.

Will remove.

tejlmand commented on samples/tfm/tfm_secure_peripheral/CMakeLists.txt:
```diff
@@ -0,0 +1,36 @@
+
+find_package(Zephyr REQUIRED HINTS $ENV{ZEPHYR_BASE})
+
+get_target_property(TFM_BINARY_DIR tfm TFM_BINARY_DIR)
```
why are you fetching `TFM_BINARY_DIR` here ?
Did you intend to use it below instead of `CMAKE_CURRENT_BINARY_DIR` but changed your mind.

joerchan replied:
configure_file below uses it in the yaml.in file.

tejlmand reviewed (changes requested):
some minor observations

mia-ko commented on samples/tfm/tfm_secure_peripheral/README.rst:
```diff
@@ -0,0 +1,166 @@
+Configuration
+*************
+
+In order to configure a peripheral as secure the peripheral must be enabled and assigned to the partition, and any interrupt signels must be defined.
```
```suggestion
In order to configure a peripheral as secure, the peripheral must be enabled and assigned to the partition, and any interrupt signals must be defined.
```

mia-ko commented on samples/tfm/tfm_secure_peripheral/README.rst:
```diff
@@ -0,0 +1,166 @@
+
+In order to configure a peripheral as secure the peripheral must be enabled and assigned to the partition, and any interrupt signels must be defined.
+
+Secure Peripheral
+=================
+
+In order to start using a peripheral as a secure peripheral it must first be enable to be used in the secure domain.
```
```suggestion
Secure peripheral
=================

To start using a peripheral as a secure peripheral, it must first be enabled for use in the secure domain.
```

mia-ko commented on samples/tfm/tfm_secure_peripheral/README.rst:
```diff
@@ -0,0 +1,166 @@
+        CONFIG_NRF_SPIM3_SECURE=y
+        CONFIG_NRF_GPIOTE0_SECURE=y
+
+In order to use GPIO pins or DPPI channels with secure peripherals they must be assigned as secure pins or channels
+This is done with a bitmask, for example to assign GPIO pin 23 as secure:
+
+.. code-block:: none
+
+	CONFIG_NRF_GPIO0_PIN_MASK_SECURE=0x00800000
+
+In addition the peripheral must be assigned to the specific partition in the partition manifest file:
```
```suggestion
To use GPIO pins or DPPI channels with secure peripherals, assign them as secure pins or channels.
This is done with a bitmask, for example to assign GPIO pin 23 as secure:

.. code-block:: none

	CONFIG_NRF_GPIO0_PIN_MASK_SECURE=0x00800000

In addition, the peripheral must be assigned to the specific partition in the partition manifest file:
```

mia-ko commented on samples/tfm/tfm_secure_peripheral/README.rst:
```diff
@@ -0,0 +1,166 @@
+                },
+        ]
+
+Secure Interrupt
```
```suggestion
Secure interrupt
```

mia-ko commented on samples/tfm/tfm_secure_peripheral/README.rst:
```diff
@@ -0,0 +1,166 @@
+Secure Interrupt
+================
+
+If the secure peripheral generates interrupts the interrupt needs to be integrated with TF-Ms interrupt mechanism.
```
```suggestion
If the secure peripheral generates interrupts, the interrupt needs to be integrated with TF-Ms interrupt mechanism.
```

mia-ko commented on samples/tfm/tfm_secure_peripheral/README.rst:
```diff
@@ -0,0 +1,166 @@
+                },
+        ]
+
+If the type of interrupt handling is defined as First Level Interrupt Handling (FLIH) then the FLIH handler needs to be defined.
```
```suggestion
If the type of interrupt handling is defined as the First Level Interrupt Handling (FLIH), then the FLIH handler needs to be defined.
```

mia-ko commented on samples/tfm/tfm_secure_peripheral/README.rst:
```diff
@@ -0,0 +1,166 @@
+                }
+        }
+
+If the type of interrupt handling is defined as Second Level Interrupt Handling (SLIH) or the FLIH handler specifies the signal to be raised the interrupt is sent to the partition as a signal.
+The way to clear a interrupt signal depends on the type of interrupt handling.
```
```suggestion
If the type of interrupt handling is defined as the Second Level Interrupt Handling (SLIH), or the FLIH handler specifies the signal to be raised, the interrupt is sent to the partition as a signal.
How to clear an interrupt signal depends on the type of interrupt handling.
```

mia-ko commented on samples/tfm/tfm_secure_peripheral/README.rst:
```diff
@@ -0,0 +1,166 @@
+
+.. note::
+
+        TF-M interrupt signals only asserts the signal but does not schedule the partition to run.
+        In cases where the interrupt signal is preempting the non-secure execution the interrupt signal will not be processed until next time the partition is scheduled to run.
+        The sample demonstrates a workaround for this limitation by triggering an EGU0 interrupt in the non-secure firmware which calls the secure partition to process the interrupt signals.
```
```suggestion
        TF-M interrupt signals only assert the signal but do not schedule the partition to run.
        In cases where the interrupt signal is preempting the non-secure execution, the interrupt signal is not processed until the next time the partition is scheduled to run.
        The sample demonstrates a workaround for this limitation by triggering an EGU0 interrupt in the non-secure firmware which calls the secure partition to process the interrupt signals.
```

mia-ko commented on samples/tfm/tfm_secure_peripheral/README.rst:
```diff
@@ -0,0 +1,166 @@
+        IRQ: GPIOTE0 count: 0x00000003
+        IRQ: TIMER1 count: 0x00000002
+
+In addition the sample sends a hash as text output using SPI on two GPIO pins.
+Connect a logic analyzer to the pins that are used for SPI output.
```
```suggestion
In addition, the sample sends a hash as a text output using SPI on two GPIO pins.
Connect a logic analyzer to the pins that are used for the SPI output.
```

joerchan commented on samples/tfm/tfm_secure_peripheral/README.rst:
```diff
@@ -0,0 +1,166 @@
+   :header: heading
+   :rows: nrf5340dk_nrf5340_cpuapp_ns, nrf9160dk_nrf9160_ns
+
+The sample also requires a logic analyzer.
```
No, anything that can observe the debug pins. We use this wording in other places, I re-used this sentence.

joerchan commented on samples/tfm/tfm_secure_peripheral/README.rst:
```diff
@@ -0,0 +1,166 @@
+It contains the partition sources, build files and build configuration files.
+The partision is built by the TF-M build system, refer to :ref:`tfm_build_system` for more details.
+
+For more information on how to add custom secure partitions refer to TF-M's guide: https://tf-m-user-guide.trustedfirmware.org/docs/integration_guide/services/tfm_secure_partition_addition.html
```
that is the same page, but hosted by NCS instead, right? Then yes we should use that URL instead. 

Will add it to links.txt

mia-ko commented on samples/tfm/tfm_secure_peripheral/README.rst:
```diff
@@ -0,0 +1,166 @@
+*************
+
+In order to configure a peripheral as secure, the peripheral must be enabled and assigned to the partition, and any interrupt signals must be defined.
+
```
```suggestion

|config|

```
We always use the stock phrase that is included with |config| in the Configuration section.

This adds the following text:
See [Configuring your application](https://developer.nordicsemi.com/nRF_Connect_SDK/doc/latest/nrf/gs_modifying.html#configure-application) for information about how to permanently or temporarily change the configuration.

mia-ko reviewed (changes requested):
Just a few things now. Are we planning to add any images btw? Not required, I just ask to see if I should set aside some resources for that :)

anangl commented on samples/tfm/tfm_secure_peripheral/secure_peripheral_partition/secure_peripheral_partition.c:
```diff
@@ -0,0 +1,252 @@
+static uint32_t m_button_count;
+static uint32_t m_timer_count;
+static uint32_t m_trigger_count;
+static uint8_t m_gpiote_channel;
```
This one is in fact used as a constant, so you could use `#define` instead.

anangl commented on samples/tfm/tfm_secure_peripheral/secure_peripheral_partition/secure_peripheral_partition.c:
```diff
@@ -0,0 +1,252 @@
+static void spim_init(uint32_t sck_pin, uint32_t mosi_pin)
+{
+	nrf_spim_pins_set(NRF_SPIM3, sck_pin, mosi_pin, NRF_SPIM_PIN_NOT_CONNECTED);
+	nrf_spim_configure(NRF_SPIM3, NRF_SPIM_MODE_0, SPIM_CONFIG_ORDER_MsbFirst);
```
```suggestion
	nrf_spim_configure(NRF_SPIM3, NRF_SPIM_MODE_0, NRF_SPIM_BIT_ORDER_MSB_FIRST);
```

hakonfam commented on subsys/nonsecure/Kconfig:
```diff
@@ -0,0 +1,576 @@
+	  been configured as a secure peripheral then only the secure processing
+	  environment is able to access the peripheral.
+
+config NRF_REGULATORS_SECURE
+	bool "REGULATORS configured as secure peripheral"
+	depends on HAS_HW_NRF_REGULATORS
+	help
+	  REGULATORS is implemented with user-selectable security mapping. If it has
+	  been configured as a secure peripheral then only the secure processing
+	  environment is able to access the peripheral.
```
Most of these look identical except the name of the peripheral, could we use a template here?

```
peripheral=REGULATORS
rsource "Kconfig.template.peripheral
```

```
# Kconfig.template.peripheral
config NRF_$(peripheral)_SECURE
    bool "$(peripheral) configured as secure peripheral"
    depends on HAS_HW_NRF_$(peripheral)
    help
      $(peripheral) is implemented with user ...
  ```

joerchan replied:
Good suggestion, thank you.

FPU and CTRLAP and GPIOTE1 did not match 100%, so kept them as they were.

hakonfam commented on subsys/spm/Kconfig:
```diff
@@ -168,237 +169,28 @@ config SPM_BLOCK_NON_SECURE_RESET
-	depends on !SPM_NRF_RTC1_NS
-	help
-	  RTC1 cannot be used the system clock driver for SPM if it is Non-Secure
+	# RTC1 cannot be used the system clock driver for SPM if it is not Secure
```
Why is this moved from help text to a comment?

joerchan replied:
Will revert.

hakonfam commented on subsys/spm/Kconfig:
```diff
@@ -168,237 +169,28 @@ config SPM_BLOCK_NON_SECURE_RESET
-config SPM_NRF_RESET_NS
-	bool "Reset is Non-Secure"
-	default y
+	# Use the Cortex-M SYSTICK timer as the system clock driver since RTC1 is Non-Secure
```
Why is this moved from help text to a comment?

hakonfam commented on subsys/spm/spm.c:
```diff
@@ -288,180 +288,175 @@ static int spm_config_unlocked_peripheral(uint8_t id, bool dma_present)
 #endif
 #ifdef NRF_CLOCK
-		PERIPH("NRF_CLOCK", NRF_CLOCK, CONFIG_SPM_NRF_CLOCK_NS),
+		PERIPH("NRF_CLOCK", NRF_CLOCK, CONFIG_NRF_CLOCK_SECURE),
```
Its almost hard to look at this not being pulled out into macros

```
/* Or whatever the syntax would be */
#define CFG_PERIPH(periph) PERIPH("NRF_#periph#", NRF_#periph#, CONFIG_NRF_#periph#_SECURE),
/* ... */
CFG_PERIPH(CLOCK)
CFG_PERIPH(RTC0)
/* ... */
```

But I guess we won't invest much in spm.c anymore, so don't care about this comment.

joerchan replied:
Not worth spending time on, and there are complications where NRFX and MDK don't have the same name, e.g. USBREG - USBREGULATORS

hakonfam commented on subsys/spm/spm.c:
```diff
@@ -472,33 +467,29 @@ static void spm_config_peripherals(void)
-		NRF_SPU->GPIOPORT[1].PERM = 0;
-	}
+#if defined(CONFIG_NRF_GPIO1_PIN_MASK_SECURE)
+	NRF_SPU->GPIOPORT[1].PERM = CONFIG_NRF_GPIO1_PIN_MASK_SECURE;
```
AS commented above, you should use HAL calls here as well.

Why did you change from `IS_ENABLED` ?

joerchan replied:
The change is from:
 ```
 #ifdef CONFIG_SOC_NRF5340_CPUAPP
     if (IS_ENABLED(CONFIG_SPM_NRF_P1_NS)) {
 ```
 to
 ```
 #if defined(CONFIG_NRF_GPIO1_PIN_MASK_SECURE)
 ```
 Can't use `CONFIG_NRF_GPIO1_PIN_MASK_SECURE` in an IS_ENABLED when it is not defined.

hakonfam commented on subsys/spm/spm.c:
```diff
@@ -472,33 +467,29 @@ static void spm_config_peripherals(void)
 #ifndef CONFIG_SPM_BOOT_SILENTLY
 		PRINT("%02u %-21s%s", i, periph[i].name,
-		      periph[i].nonsecure ? "Non-Secure" : "Secure\t");
+		      periph[i].secure ? "Secure\t" : "Non-Secure");
```
If you check the inverse you will save testers time rewriting their regexes used for testing.

`!secure ? "Non-secure": "Secure"`

joerchan replied:
Unless I'm missing something the output is not affected by this change. the `\t` is in the same place.

hakonfam replied:
Ah, I read it incorrectly.

hakonfam commented on modules/tfm/tfm/boards/board/tfm_peripherals_config.h:
```diff
@@ -11,18 +11,158 @@
+#endif
 
-#define TFM_PERIPHERAL_WDT0_SECURE 1
+/* TODO: Update to TFM_PERIPHERAL_DPPI_CHANNEL_SECURE_MASK once this is renamed in TF-M. */
```
Rename what to that? Isn't it already called that here?

joerchan replied:
Comment is wrong, should be `TFM_PERIPHERAL_DPPI_CHANNEL_MASK_SECURE`

hakonfam commented on samples/tfm/tfm_secure_peripheral/README.rst:
```diff
@@ -0,0 +1,166 @@
+
+.. include:: /includes/build_and_run.txt
+
+Testing
```
Why is the logic analyzer needed? Can't you verify the behavior by looking at this console output?

joerchan replied:
It's not required, the sample demonstrates using SPI in secure. A logic analyzer can be used to see some pin toggling of "random" data.

Will have to check with the testers how best to proceed with this.

hakonfam commented on samples/tfm/tfm_secure_peripheral/src/secure_peripheral_partition.c:
```diff
@@ -0,0 +1,33 @@
+/*
+ * Copyright (c) 2021 Nordic Semiconductor ASA.
+ *
+ * SPDX-License-Identifier: Apache-2.0
```
Is the year and license id correct?

hakonfam reviewed (commented):
Sorry for the late review, won't block anything.

mbolivar-nordic reviewed (approved):
ACKed for now following offline discussion

joerchan commented on samples/tfm/tfm_secure_peripheral/secure_peripheral_partition/tfm_secure_peripheral_partition.yaml:
```diff
@@ -0,0 +1,77 @@
+#-------------------------------------------------------------------------------
+# Copyright (c) 2018-2020, Arm Limited. All rights reserved.
+# Copyright (c) 2021, Nordic Semiconductor ASA. All rights reserved.
+#
```
No, will fix.

joerchan commented on samples/tfm/tfm_secure_peripheral/secure_peripheral_partition/secure_peripheral_partition.c:
```diff
@@ -0,0 +1,252 @@
+#include "util.h"
+
+#if defined(CONFIG_BOARD_NRF9160DK_NRF9160_NS)
+#define BUTTON_PIN        (6) /* P0.06 */
```
Button0 is already in DTS, added SPI pins in overlay.

anangl commented on subsys/nonsecure/Kconfig:
```diff
@@ -0,0 +1,262 @@
+          If it has been configured as a secure peripheral then only the secure
+          processing environment is able to access the peripheral
+
+peripheral=NRF_DCNF_SECURE
+rsource "Kconfig.template.peripheral_secure"
```
I guess this was supposed to be:
```suggestion
peripheral=DCNF
rsource "Kconfig.template.peripheral_secure"
```
And similarly for a few next entries.

anangl commented on subsys/nonsecure/Kconfig:
```diff
@@ -0,0 +1,273 @@
+peripheral=RTC1
+rsource "Kconfig.template.peripheral_secure"
+
+config NRF_DPPI_SECURE
```
Why not use the template also here? The peripheral is in fact named DPPIC, so `NRF_DPPIC_SECURE` would be an appropriate option for it.

joerchan replied:
Well, since both are being used I prefer to use the one that introduces the least amount of naming differences to the user. And that means aligning them to the nrfx naming. For example:
```
config NRFX_DPPI
	bool "Enable DPPI allocator"
	depends on HAS_HW_NRF_DPPIC
```

anangl replied:
When it comes to the peripheral itself, only DPPIC is used. That's why the `HAS_HW_NRF_DPPIC` option has this C at the end and why there is a `dppic` node in devicetree, not `dppi`. The above option enables compilation of the allocator for DPPI channels and groups, hence it uses DPPI, not DPPIC.
But it's actually a detail, so I won't insist on this. If you prefer to keep it as it is now, I'm fine with that.

joerchan replied:
I prefer to keep it, it's not only my preference though, since I've already decided to use NRF_DPPI_ in TF-M changing it now would cause an inconsistency there as well.

NordicBuilder commented:
### Integration test specification

| Test Module | File based changes | Manually selected | West overwrite |
| ----------- | :----------------: | :---------------: | :------------: |
| test-ci-nrfconnect-boot-fw-update | X | | |
| test-fw-nrfconnect-chip | X | | |
| test-fw-nrfconnect-tfm | X | | |

**test-fw-nrfconnect-tfm:** added because there was no .github/test-spec.yml in 'trusted-firmware-m'

[Detailed information of selected test modules](https://jenkins-ncs.nordicsemi.no/job/latest/job/sub/job/test-fw-nrfconnect-nrf/job/master/22211/artifact/test-spec-summary.yml/*view*/)

*Note: This message is automatically posted and updated by the CI*

anangl commented on subsys/nonsecure/Kconfig:
```diff
@@ -0,0 +1,273 @@
+	  made to GPIO pins configured as non-secure.
+
+config NRF_GPIO1_PIN_MASK_SECURE
+	hex "Set GPIO0 PERM mask for Secure"
```
```suggestion
	hex "Set GPIO1 PERM mask for Secure"
```

anangl commented on subsys/spm/spm.c:
```diff
@@ -105,8 +105,8 @@ extern int irq_target_state_is_secure(unsigned int irq);
  */
-#if (CONFIG_SPM_NRF_DPPIC_PERM_MASK >= (1 << DPPI_CH_NUM))
-#error "SPM_NRF_DPPIC_PERM_MASK exceeds number of available DPPI channels"
+#if (CONFIG_NRF_DPPI_MASK_SECURE >= (1 << DPPI_CH_NUM))
+#error "CONFIG_NRF_DPPI_MASK_SECURE exceeds number of available DPPI channels"
```
It should be `CONFIG_NRF_DPPI_CHANNEL_MASK_SECURE`.

mbolivar-nordic commented:
@joerchan can you confirm you've addressed @tejlmand 's comments? I've dismissed his review for now since he is on leave, and I guess we can merge this if you rebase and address the merge conflict if so.

joerchan commented:
> joerchan can you confirm you've addressed tejlmand 's comments? I've dismissed his review for now since he is on leave, and I guess we can merge this if you rebase and address the merge conflict if so.

Yes, I have addressed his comments. Rebased and fixed merge conflicts.

mbolivar-nordic commented:
@joerchan I re-ran CI, and it's green now. You should be all set to merge if you can get another +1 on https://github.com/nrfconnect/sdk-trusted-firmware-m/pull/45.

rlubos commented on west.yml:
```diff
@@ -113,7 +113,7 @@ manifest:
       repo-path: sdk-trusted-firmware-m
       path: modules/tee/tf-m/trusted-firmware-m
-      revision: e6f3c85f618e51b6b8fb0105e2ce9fa6372ba8f0
+      revision: pull/45/head
```
Please update SHA: `7d4a7139009c24ffa6edfb7f24fe32090bca0583`

joerchan replied:
Done.

lemrey commented:
Just a question, have not been following TFM so perhaps the answer is obvious, but why are we inverting the logic? Earlier we specified which peripherals were non-secure because secure is the hardware default. Now instead, we specify which peripherals are secure and configure everything else as non-secure.

joerchan commented:
@lemrey Because configs that are `default n` are easier to deal with than `default y`. You can do select and imply on them.
